[ENH] Univariate time series bootstrapping

ltsaprounis opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
Implements the `boxcox -> STL -> remainder moving block bootstraping` algorithm described in [Forecasting Principles and Practice - Chapter 12.5](https://otexts.com/fpp3/bootstrap.html). The transformer takes as input a series and returns a panel of similar time series. The panel can then be used to create bagged forecasts and prediction intervals. Will implement the `BaggedForecaster` after this is merged.

PS: thanks @aiwalter for the STL parameter docstrings, I shamelessly stole these from `STLForecaster` 😄 

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

1. The transformer takes as input a series and outputs a multi-index Panel mtype. Need a second check on the tags to make sure they're ok. 

#### Example usage
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

```python
from sktime.transformations.bootstrap import STLBootsrapTransformer
from sktime.datasets import load_airline
from sktime.utils.plotting import plot_series

ts = load_airline()

t = STLBootsrapTransformer(10)
df = t.fit_transform(ts)

series_list = []
names = []
for group, series in df.groupby(level=[0], as_index=False):
    series.index = series.index.droplevel(0)
    series_list.append(series)
    names.append(group)

plot_series(
    *series_list,
    labels=names
)
plt.show()
```
![image](https://user-images.githubusercontent.com/64217214/153780658-40f3570b-577a-4a33-93c8-4637ad626dd5.png)


#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [X] I've added unit tests and made sure they pass locally.

##### For new estimators
- [X] I've added the estimator to the online documentation.



<!--
Thanks for contributing!
-->

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,330 @@
+            )
+
+        # fit boxcox to get lambda and transform X
+        self.BoxCoxTransformer_ = BoxCoxTransformer(
```
This should be snake case I think

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+    similar time series to the input time series [1]_, [2]_.
+
+    First the observed time series is transformed using a Box-Cox transformation to
+    stabilise the variance. Then it's decomposed to seasonal, trend and remainder
```
```suggestion
    stabilise the variance. Then it's decomposed to seasonal, trend and residual
```

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+
+    First the observed time series is transformed using a Box-Cox transformation to
+    stabilise the variance. Then it's decomposed to seasonal, trend and remainder
+    time series, using the STL algorithm from statsmodels [4]_. We then sample blocks
```
```suggestion
    time series, using the STL implementation from statsmodels [4]_. We then sample blocks
```

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+    of the observed series. Finally these bootstrapped remainders are added to the
+    season and trend components and we use the inverse Box-Cox transform to return
+    a panel of similar time series.
+
```
Could you maybe describe here also some examples of good use-cases for this transformer?

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+
+    Parameters
+    ----------
+    number_of_new_series : int, optional
```
```suggestion
    n_series : int, optional
```

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+        If provided, the synthetic series names will have the series_name as a
+        prefix followed by an undescore e.g. "<series_name>_synthetic_1". If not
+        provided the series names will no prefix will be present e.g. "synthetic_1".
+    boxcox_bounds : Tuple, optional
+        Lower and upper bounds used to restrict the feasible range
+        when solving for the value of lambda, by default None.
+    boxcox_method : str, optional
```
```suggestion
    bounds : Tuple, optional
        Lower and upper bounds used to restrict the feasible range
        when solving for the value of lambda, by default None.
    method : str, optional
```
I would keep original bixcox param names and rather describe in docstring that it goes into boxcox

ltsaprounis replied:
changed it to `lambda_bounds` & `lambda_method`

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+from sktime.transformations.series.boxcox import BoxCoxTransformer
+
+
+class UnivariateBootsrappingTransformer(BaseTransformer):
```
this is a long class name :) what about just `BootstrapTransformer`? The univariate in the name is at least not needed imho.

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+    ... )
+    >>> from sktime.datasets import load_airline
+    >>> from sktime.utils.plotting import plot_series
+    >>> ts = load_airline()
+    >>> ubt = UnivariateBootsrappingTransformer(10)
+    >>> df = ubt.fit_transform(ts)
```
```suggestion
    >>> y = load_airline()
    >>> transformer = UnivariateBootsrappingTransformer(10)
    >>> y_hat = transformer.fit_transform(y)
```

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+    >>> df = ubt.fit_transform(ts)
+    >>> series_list = []
+    >>> names = []
+    >>> for group, series in df.groupby(level=[0], as_index=False):
```
```suggestion
    >>> for group, series in y_hat.groupby(level=[0], as_index=False):
```

aiwalter commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,395 @@
+
+        return df
+
+    @staticmethod
+    def moving_block_bootstrap(
```
```suggestion
    def _moving_block_bootstrap(
```

ltsaprounis replied:
keeping it static, but making it private

fkiraly commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,401 @@
+    """Creates a population of similar time series.
+
+    This method utilises a form of bootstrapping to generate a population of
+    similar time series to the input time series [1]_, [2]_.
```
the references are textbooks, can you kindly point to the section/chapter this is specifically implementing?

fkiraly commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,401 @@
+
+    First the observed time series is transformed using a Box-Cox transformation to
+    stabilise the variance. Then it's decomposed to seasonal, trend and residual
+    time series, using the STL implementation from statsmodels [4]_. We then sample
```
please point to the class/function/method

fkiraly commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,401 @@
+    return a panel of similar time series.
+
+    The resulting panel can be used for Bagging forecasts, prediction intervals and
+    data augmentation.
```
the description should detail the structure of the returned data frame or panel object, since multiple variables are generated. Since the panel object could be `numpy`, the order of columns should also figure in this description.

fkiraly commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,401 @@
+        Must be an odd integer. If not provided uses the smallest odd integer greater
+        than 1.5 * period / (1 - 1.5 / seasonal), following the suggestion in the
+        original implementation.
+    low_pass : int, optional
```
this should say that this parameter belongs to STL

fkiraly commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,401 @@
+    The resulting panel can be used for Bagging forecasts, prediction intervals and
+    data augmentation.
+
+    Parameters
```
perhaps order parameters by component, and/or say which component they belong to.

fkiraly commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,401 @@
+    """Creates a population of similar time series.
+
+    This method utilises a form of bootstrapping to generate a population of
+    similar time series to the input time series [1]_, [2]_.
```
Hyndman (I assume section 12.5?) does not mention the BoxCox transformer explicitly.
Personally, I would interpret the text that the bootstrap is STL plus the mentioned block bootstrap method.

ltsaprounis replied:
Yes it's a bit vague in the 3rd edition, but the original reference is Bergmeir 2016. In the second edition of forecasting principles and practice, Hyndman mentions BoxCox explicitly ([fpp 2nd edition chapter 11.4](https://otexts.com/fpp2/bootstrap.html)). In general all the implementations and papers I've seen that utilise this bootstrapping method always have BoxCox (with Guerrero to find lambda) as the first step. 

Conceptually this makes sense. The variance needs to be stable for MBB on the residuals to work. The transformed series has stable variance and when the inverse transform is applied in the end, the synthetic series has the same variance pattern as the original series.

fkiraly replied:
well, you could also say: Bergmeir is a pipeline [boxcox][bootstrap][bagging], and this estimator should be [bootstrap]
Or, you could say, Bergmeir is [boxcox+bootstrap][bagging], and this estimator should be [boxcox+bootstrap].

If there is no computational advantage to put boxcox and the bootstrap together, I´d personally split them up, and then use a pipeline constrution when implementing the full Bergmeir.

But I have no strong opinion on this. Doing the "combined" version just makes the docstrings more complicated, and may result in us having to implement the pure [bootstrap] separately later (time series bootstrap resampling is generally a common component).

fkiraly commented on sktime/transformations/series/bootstrapping.py:
```diff
@@ -0,0 +1,401 @@
+        If True the output will contain the actual time series, by default True.
+        The actual time series will be labelled as "<series_name>_actual" (or "actual"
+        if series name is None).
+    series_name : str, optional
```
should this not be passed as the column name in a single column `pd.DataFrame`, or name of a `pd.Series`, instead of a parameter?

fkiraly reviewed (changes requested):
This looks good!

I have questions that should be clarified (relevant? obsolete?) or acted upon (fix):
* a bundle of improvement comments about the docstring
* is the box-cox transform really part of this estimator, according to Hyndman?
* would it be worth implementing the block bootstrap in isolation, too? All the methods are there.
* is the `series` folder the right folder here? Or should it be in a `resample` folder or `series_to_panel` folder, or similar? Together with augmentation techniques?

fkiraly commented:
PS: just saw that the second ref, Bergmeir 2016, includes boxcox - but it also includes bagging

ltsaprounis commented:
> PS: just saw that the second ref, Bergmeir 2016, includes boxcox - but it also includes bagging

Yes! we need this transformer first and then I'll implement the `BaggingForecaster`. 
`BaggingForecaster` will generate a point forecast for each series (using any forecaster) and we can then use the mean as the point forecast and the quantiles from the simulated series as the prediction intervals. 

I also have my eye on the [`tfp.distributions.Empirical`](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Empirical) as the probabilistic interface for this kind of sampling methods. Another method where the same logic can be used is this one: https://otexts.com/fpp3/prediction-intervals.html#prediction-intervals-from-bootstrapped-residuals

ltsaprounis commented:
> This looks good!
> 
> I have questions that should be clarified (relevant? obsolete?) or acted upon (fix):
> 
> * a bundle of improvement comments about the docstring
> * is the box-cox transform really part of this estimator, according to Hyndman?

I'd say yes! (see detailed explanation in the comment section above)
> * would it be worth implementing the block bootstrap in isolation, too? All the methods are there.

Good idea! will do it in this PR
> * is the `series` folder the right folder here? Or should it be in a `resample` folder or `series_to_panel` folder, or similar? Together with augmentation techniques?

Agreed! What name should we go for? `series_to_panel`?

fkiraly commented:
> I also have my eye on the [`tfp.distributions.Empirical`](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Empirical) as the probabilistic interface for this kind of sampling methods.

Yes, exactly! One of my minor points that tipped my personal scales towards `tfp` was performant support for empirical distribiutions.

ltsaprounis commented:
Changes to the PR: 
1. Renamed original Transformer to `STLResidualBootstrapTransformer`, added the original `MovingBlockBootstrapTransformer`. 
2. Moved bootstrapping Transformers to `sktime.transformations.bootstrap`
3. Addressed docstring changes from @fkiraly

fkiraly commented:
uh-oh, tests are failing!

`X length in STLResidualBootsrapTransformer should be greater than block_length`

indicates at parameters that are not applicable to the test data which has time series of length 10.
Probably `block_length` needs to be smaller than 10?

ltsaprounis commented:
> uh-oh, tests are failing!
> 
> `X length in STLResidualBootsrapTransformer should be greater than block_length`
> 
> indicates at parameters that are not applicable to the test data which has time series of length 10. Probably `block_length` needs to be smaller than 10?
> 
> (I wouldn´t change the default, just the test parameter setting in `get_test_params`)

Yes! thanks for the tip. Took me a while to figure it out.

`test_fit_idempotent` is still failing for the `MovingBlockBootsrapTransformer`, I'm guessing it's because of the random nature of the transform. @fkiraly do you have any idea on how to deal with this?

Will also write a few more tests to make sure all the exceptions are tested.

aiwalter commented:
without looking into all discussions, why is is named `STLResidualBootsrapTransformer`? `STL` is per definition returning residuals, so we might rather go for `STLBootsrapTransformer`?

ltsaprounis commented:
> without looking into all discussions, why is is named `STLResidualBootsrapTransformer`? `STL` is per definition returning residuals, so we might rather go for `STLBootsrapTransformer`?

I'll be the first to admit that I'm bad with names 😄 
My thought process was that STL returns trend, seasonal and residual series and the Transform bootstraps the residuals while keeping the other 2 as is. If you think `STLBootstrapTranform` is informative enough I prefer it too as it's a bit shorter.

ltsaprounis commented:
@fkiraly / @aiwalter 
Changed name to STLForecaster and fixed the `test_fit_idempotent` by adding the `random_state` parameter.
Good to review again!

fkiraly commented on docs/source/api_reference/transformations.rst:
```diff
@@ -360,3 +360,25 @@ FeatureSelection
 
     FeatureSelection
+
+STLResidualBootsrapTransformer
```
typo, should be Bootstrap

fkiraly commented on sktime/transformations/bootstrap/__init__.py:
```diff
@@ -0,0 +1,9 @@
+# -*- coding: utf-8 -*-
+"""Bootstrapping methods for time series."""
```
perhaps explain here, and in the docs, that this takes a single series and produces a panel of series?

fkiraly reviewed (dismissed):
Looks good to me now!

Let's wait for @aiwalter to check whether his concerns were addressed.

Meanwhile, I've added some minor non-blocking doc comments.

ltsaprounis commented:
@aiwalter - Should be good to go now, can you have a look and review/approve?

fkiraly commented:
merging to unblock further work of @ltsaprounis.
No review activity since one week, and all change requests by @aiwalter seemed resolved.
@aiwalter, let us know if you had any major reservations in relation to your change requests, and feel free to unmerge.
